File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/GameSeries.java
package com.hiddenramblings.tagmo.amiibo;

import androidx.annotation.NonNull;

public class GameSeries implements Comparable<GameSeries> {
    public static final long MASK = 0xFFC0000000000000L;
    public static final int BITSHIFT = 4 * 13;

    public final AmiiboManager manager;
    public final long id;
    public final String name;

    public GameSeries(AmiiboManager manager, long id, String name) {
        this.manager = manager;
        this.id = id;
        this.name = name;
    }

    public GameSeries(AmiiboManager manager, String id, String name) {
        this(manager, hexToId(id), name);
    }

    public static long hexToId(String value) {
        return (Long.decode(value) << BITSHIFT) & MASK;
    }

    @Override
    public int compareTo(@NonNull GameSeries gameSeries) {
        return this.name.compareTo(gameSeries.name);
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/amiibo/GameSeries.kt
package com.hiddenramblings.tagmo.amiibo

class GameSeries(val manager: AmiiboManager, val id: Long, val name: String) :
    Comparable<GameSeries> {
    constructor(manager: AmiiboManager, id: String, name: String) : this(manager, hexToId(id), name)

    override fun compareTo(other: GameSeries): Int {
        return name.compareTo(other.name)
    }

    companion object {
        const val MASK = -0x40000000000000L
        const val BITSHIFT = 4 * 13
        fun hexToId(value: String): Long {
            return java.lang.Long.decode(value) shl BITSHIFT and MASK
        }
    }
}